### PR TITLE
General improvements, changes, and fixes

### PR DESCRIPTION
**Pull request by Myned**
These commits add functionality and quality of life improvements without majorly changing how the executable section operates, but most setups will need to modify and re-add their executables to align with regex syntax. The changes may also negate the need for the warning about running multiple exes simultaneously.

This pull is fully functional on my system, so it seems merge-ready, but there might be some interactions I have missed.

_Tested on Manjaro KDE 5.18.5, Qt 5.14.2_

### CODE REVIEW

**Comment by WarmUpTill** [comment]
Looks good!

I am just not sure about moving the advanced-scene-switcher.ui to the headers folder.
At least on my windows machine it seems to cause issues with in-tree builds.

I guess this behaviour differs for the different platforms or do you also encounter the same issue when building in-tree?

**Comment by Myned** [comment]
After testing an in-tree build on the same system, the error does not seem to occur, so it is limited to out-of-tree builds only. I don't know enough about the CMake process to pinpoint where the discrepancy occurs, but here is the traceback:

```
[  5%] Automatic MOC and UIC for target advanced-scene-switcher

AutoUic error
-------------
"SRC:/src/headers/advanced-scene-switcher.hpp"
includes the uic file "ui_advanced-scene-switcher.h",
but the user interface file "advanced-scene-switcher.ui"
could not be found in the following directories
"SRC:/src/headers"

make[2]: *** [CMakeFiles/advanced-scene-switcher_autogen.dir/build.make:78: CMakeFiles/advanced-scene-switcher_autogen] Error 1
make[1]: *** [CMakeFiles/Makefile2:124: CMakeFiles/advanced-scene-switcher_autogen.dir/all] Error 2
make: *** [Makefile:150: all] Error 2
```

If needed, I can submit an issue instead.

**Comment by Myned** [comment]
I would also hold off a bit on merging as I am currently working on adding the ability to ignore windows (using the existing section under Window Title) for the executable switches, and fixing a small bug with it not ignoring when fullscreen/focus is checked.

**Comment by Myned** [comment]
The fix for fullscreen not respecting the ignored windows is proving to be more difficult than I first thought. The problem is that the [`isFullscreen()`](https://github.com/WarmUpTill/SceneSwitcher/blob/228a18eb77aa1a24048d992ed7142be6d891e8e2/src/linux/advanced-scene-switcher-nix.cpp#L243) function uses the active window to check if it is fullscreen, not one that is sent as an argument, so if you tab to another (ignored) window, the check returns false and the match fails. The executable switch works, though, since it relies on the [`isInFocus(...)`](https://github.com/WarmUpTill/SceneSwitcher/blob/228a18eb77aa1a24048d992ed7142be6d891e8e2/src/linux/advanced-scene-switcher-nix.cpp#L293) check, which requires the switch as an argument.

I would fix the function myself, but it is OS-bound and even just the Linux process is daunting, to say the least. Speaking of which, I neglected to realize that the check for focus is also OS-bound, so only the Linux `isInFocus()` function works with the regex change, currently.

**Comment by WarmUpTill** [comment]
Thanks for the update!

Regarding "Remove direct match from window switcher and reorganize [...] the core logic shouldn't prevent a user from utilizing regex to its full potential":
I agree and this is how it was originally implemented. However most users won't understand what regular expressions are and will have much trouble understanding how to use them and will not understand how to escape characters.

For example in the past users had issues with the games "Rocket League (32-bit, DX9)" or "DeadByDaylight (64-bit, PCD3D_SM5)", which caused me to add this direct match functionality taking priority over regex.

Maybe I should have handled it differently (adding a checkbox to specifically select whether a entered window name should be interpreted as a regex e.g), but I would not recommend reverting this change as the use of brackets is quite common in window names unfortunately.
(I should have added a comment about that in the code, sorry)

**Comment by Myned** [comment]
Understood. Thank you for taking the time to explain that. I'll revert the change and add it to the executable switch as well, then tackle the `isInFocus()` regex on the other platforms.

**Comment by WarmUpTill** [comment]
Thanks again for the update!

Just one small thing regarding a4de8c0.
Unfortunately this will cause issues for Windows as the process name does necessarily match the window name.

So for example the window name of the process "firefox.exe" can be "Minor changes and fixes to executable switching by Myned · Pull Request #24 · WarmUpTill/SceneSwitcher - Mozilla Firefox".
In this case isInFocus() can never return true.

Thus I would suggest something like for Windows:
```
bool isInFocus(const QString &exeToCheck)
{

	// only checks if the current foreground window is from the same executable,
	// may return true for incorrent not meant windows from a program
	HWND foregroundWindow = GetForegroundWindow();
	DWORD processId = 0;
	GetWindowThreadProcessId(foregroundWindow, &processId);

	HANDLE process = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, processId);
	if (process == NULL) return false;

	WCHAR executablePath[600];
	GetModuleFileNameEx(process, 0, executablePath, 600);
	CloseHandle(process);

	QString exe = QString::fromWCharArray(executablePath).split(QRegExp("(/|\\\\)")).back();

	// True if executable switch equals current window
	bool equals = (exeToCheck == exe);

	// True if executable switch matches current window
	bool matches = exe.contains(QRegularExpression(exeToCheck));

	return (equals || matches);
}
````
I think for Mac and Linux using GetCurrentWindowTitle() might suffice as you suggested.

**Comment by Myned** [comment]
`isInFocus()` from https://github.com/WarmUpTill/SceneSwitcher/pull/24/commits/a4de8c0c590e5182035ca52a40c9329629c2adf1 uses the same method as the [window title switch](https://github.com/WarmUpTill/SceneSwitcher/blob/228a18eb77aa1a24048d992ed7142be6d891e8e2/src/window-title-switch.cpp#L229) - [`GetCurrentWindowTitle()`](https://github.com/WarmUpTill/SceneSwitcher/blob/228a18eb77aa1a24048d992ed7142be6d891e8e2/src/win/advanced-scene-switcher-win.cpp#L61), which gets the title/appname, not the process file (including on Windows, otherwise the window title switch wouldn't work).

My previous understanding was based on the assumption that the functionality was the same between platforms. However, in Linux, the original executable switch utilizes window titles and not actual executables like `firefox.exe`, unlike Windows and macOS which both use the executable/appname. So only in Windows and macOS, this means that there may be multiple matches for a single executable that don't have windows, which is why the warning was added. It also means that there would be no way for the user to choose a specific window from the same executable.

This pull and https://github.com/WarmUpTill/SceneSwitcher/pull/24/commits/a4de8c0c590e5182035ca52a40c9329629c2adf1 in particular aims to change the executable switch in Windows and macOS to be similar to a 'persistent window title switch' or the Linux executable switch in that only titles are used. This limits the amount of possible matches to only open windows, which covers the majority of usecases and standardizes the functionality differences between platforms. It also skirts around the need for determining what an executable even is (e.g. Should shell files be considered executables?).

The macOS version works with my testing for window titles because the executables that [`GetProcessList()`](https://github.com/WarmUpTill/SceneSwitcher/blob/228a18eb77aa1a24048d992ed7142be6d891e8e2/src/osx/advanced-scene-switcher-osx.mm#L111) generates are appnames like `Firefox` instead of files like `firefox.exe`. I would like to change that to use window titles, but it doesn't look like AppKit nor AppleScript have the ability to find *all* window titles (In AppleScript, for example, Telegram does not have a window tied to its process, and in AppKit, `NSRunningApplication` does not have a title property), so the executable would be largely unchanged.

For the sake of simplicity, I just wanted to change how the executable switch works, but another option would be to add a checkbox in the Window Title tab asking if the switch needs to be in focus, which would remove the need for the Executable tab and simplify the implementation.

**Comment by WarmUpTill** [comment]
I think the original intention of the executable switch was for it to be based on running processes instead of window titles.
Back when it was introduced it was only available for Windows (hence the name "**exe**cutable" switching).

I added the MacOS and Linux versions afterwards just so I could use the same UI for all platforms. As I was very unfamiliar with both of these platforms at the time I quickly gave up on implementing the exact same behavior and just settled for reusing the window title based functionality which was already working.

Today I would probably do it differently (maybe using procps / libproc for Linux).

As I am afraid of breaking stuff which was previously working for users I would prefer to not change this behavior for Windows at the moment.

I also think it would be best to align the different platform's behaviors by adjusting the executable switch for Linux, but this might be out of scope of this PR.

So for now I would suggest to merge this PR to add the support for regex for the executable switch and I will change the Windows isInFocus() to the above suggestion to avoid the issues on Windows.

Do you agree?

**Comment by Myned** [comment]
I've started working on an alternative method that aligns with that end result, so I would rather hold off on merging just yet so I can revert the title matching with the Executable switch for all platforms, including the `isInFocus()` changes for Windows.

I added a checkbox to the Window Title switch. It works as intended like the Executable switch so far, just have to fix the fullscreen bug which is near completion on Linux.

This is how I envision it to work. Let me know if it is to your liking.
screenshot

**Comment by WarmUpTill** [comment]
Yep, looks good.
Do you think it would make sense to have "only if focused" be checked by default / initialize the value to true (if it does not exist) so existing window title switches won't break?

**Comment by Myned** [comment]
When I tested it before, all of my existing switches had focus enabled, so that may not be a problem.

**Comment by Myned** [comment]
After wrapping up and retesting, it does break existing switches. Defaulting to true is probably a good idea, though I'm not sure how to accomplish that. I've focused mostly on the switching logic - the OBS initialization seems a little hard to follow at the moment. I'll push the current changes so you can take a look first.

**Comment by WarmUpTill** [comment]
Looks good and seems to be working as expected in the limited tests I performed.

Regarding the initialization you can use the following to check if the value "focus" exists:
````
bool focus_exists = obs_data_has_user_value(array_obj, "focus");
bool focus =
	obs_data_get_bool(array_obj, "focus") || ! focus_exists;
````

**Comment by Myned** [comment]
Thank you. I wanted to use `obs_data_get_default_bool()` in the [OBS documentation](https://obsproject.com/docs/reference-settings.html?highlight=obs_data#c.obs_data_get_default_bool), but it doesn't have a third argument in order to return a default value, unless there are heuristics involved of some sort.

**Comment by WarmUpTill** [comment]
Just checking in to ask if there is any way I can support you to finish up this PR :)
What are the remaining open topics?

I seem to recall that the windows isFullscreen() function was not yet working as expected last time i checked.
This would be my suggestion to resolve this.

```
bool isFullscreen(std::string &title)
{
	RECT appBounds;
	RECT rc;

	HWND hwnd = GetWindow(GetDesktopWindow(), GW_CHILD);

	while (hwnd)
	{
		string wtitle;
		if (WindowValid(hwnd) && GetWindowTitle(hwnd, wtitle) && wtitle == title)
			break;
		hwnd = GetNextWindow(hwnd, GW_HWNDNEXT);
	}

	GetWindowRect(GetDesktopWindow(), &rc);
								   
	if (hwnd != GetDesktopWindow() && hwnd != GetShellWindow())
	{
		GetWindowRect(hwnd, &appBounds);
		if (rc.bottom == appBounds.bottom && rc.top == appBounds.top && rc.left == appBounds.left
			&& rc.right == appBounds.right)
		{
			return true;
		}
	}
	return false;
}
```

**Comment by Myned** [comment]
Apologies, I was caught up in some other things.

The `isFullscreen()` function is currently not fixed in Windows and macOS, and what you have suggested is mostly what I had in mind for both. I was wondering if those platforms have something like Linux's window properties to detect fullscreen/maximization instead of using window sizes, though that's not as important outside of Linux due to users not being able to override window properties by design.

I haven't been able to set up a build environment on Windows, yet, so I can't test it there, and I'm not familiar enough with the Windows process. I have invited you to the fork if you want to change anything yourself. Otherwise, I can just add what you wrote and check any build errors from here. I can get started on the macOS fix in the meantime as I was able to set up the environment there.

I was also thinking about adding up/down arrows to be able to reorder the Window Title and Executable lists, like the Ignored list.

**Comment by WarmUpTill** [comment]
No need to apologize!

At least for windows I am not aware of a better way than just comparing the window size to the monitor size. (Although this probably causes the check to fail if the window is fullscreen on a secondary monitor, but I could never test that)
I will check if there is a better way and if there is add it to your fork.

Feel free to commit the suggestion and just check for build errors on windows.
It worked during my tests, but I will make sure to test once again the PR is complete.

I cannot test the macOS variant as OBS 25+ is no longer supported on the machine I have access to so I'm just going to have to trust you there :)

Adding the up / down arrows sounds like a good idea.

**Comment by Myned** [comment]
I found [this answer](https://stackoverflow.com/a/55542400), which seems to check the monitor that the window is assigned to, though I don't know how that considers a window that is visible on multiple monitors.

**Comment by Myned** [comment]
Also, commit 98c3131017db23b04d3185599eee0229f72ac9b2 causes a segfault on OBS startup, so I'll revert it for now. The reference to a build failure for commit f03fca85aa7960b6d9bf7205412f62ea94d141fc doesn't seem to affect my machine as well.

**Comment by WarmUpTill** [comment]
Ah, thanks for the heads up regarding the segfault!
Forgot to commit the init to null for the SwticherThread. (513fe07)
I would have much preferred to use the QThread::create but Ubuntu 18 still seems to use QT 5.9 and it was added with QT 5.10.

Thanks for the link - will test later today if it still works as expected with a single monitor and maybe try to dig up an old secondary monitor :)

(I have no clue why the last macos build failed at the point it did - I guess we can ignore that)

**Comment by Myned** [comment]
I do have another monitor to test with if need be, just need the binary for Windows if you don't mind uploading one for the fork.

Side note, which syntax do you prefer? I've been following something similar to the second one since it tends to isolate better with multiline conditions:
```cpp
void function() {
    if (true &&
        false) {
        ;
    }
}
```

```cpp
void function()
{
    if (true
        && false
    ) {
        ;
    }
}
```

**Comment by WarmUpTill** [comment]
Do you have access to the automated build artifacts?
https://github.com/WarmUpTill/SceneSwitcher/suites/767488468/artifacts/7992332
If not I will send you a link separately.

I just use https://github.com/obsproject/obs-studio/blob/master/.clang-format for automatic formatting.

**Comment by Myned** [comment]
Ohh, gotcha. I didn't know OBS had that. If you want, I can apply it to all of the `src/` files and push the `.clang-format` file as well.

I also wasn't aware of artifacts at all - the GitHub workflow process is a bit new to me. I do have access, though, thanks.

**Comment by WarmUpTill** [comment]
Sure, feel free to apply it to all files and add it to the repo. (Only have done so on files I have worked on recently so far)

**Comment by Myned** [comment]
One thing to note about the latest commit to the UI changes: in order to make the literal string `Region` work for the tab name, the `translatable` property had to be disabled. Since the rest of the plugin doesn't seem to change depending on the OBS language, I figured that was acceptable regardless.

**Comment by Myned** [comment]
The Windows fullscreen check doesn't seem to work at all on my Bootcamp install, even when the window is in focus. Using a build before commit https://github.com/WarmUpTill/SceneSwitcher/pull/24/commits/5590e97f4fe5b8956ca45799f48d470ba4f9f52f works, but only while in focus. It also doesn't switch for maximized windows, is that intentional?

**Comment by WarmUpTill** [comment]
Oh that is surprising.
I tested it with using a Firefox window in fullscreen mode (not maximized) on a single monitor setup.
I will investigate later today.

**Comment by WarmUpTill** [comment]
I pulled the most recent changes and fullscreen detection is still working as expected on windows.
Maybe it is an issue only occurring multiple monitor setups?
(Could not find a second monitor unfortunately)

If you want and still have issues with your windows build setup I can prepare a version that writes the current coordinates of the window and monitor to the OBS log, by adding the following to the fullscreen function:
``
blog(0, "app - b: %d, t: %d, l: %d, r: %d", appBounds.bottom, appBounds.top, appBounds.left, appBounds.right);
blog(0, "mon - b: %d, t: %d, l: %d, r: %d", monitorInfo.rcMonitor.bottom, monitorInfo.rcMonitor.top, monitorInfo.rcMonitor.left, monitorInfo.rcMonitor.right);
``


One thing I noted however is that the ignore window functionality does not seem to be working for window titles that are invalid regular expressions. 
I would suggest adding the direct match `title == s.window` there also:

``
		// True if current window is ignored AND switch matches last window
		bool ignore =
			(ignored &&
			 (title == s.window ||
			  QString::fromStdString(title).contains(
				  QRegularExpression(
					  QString::fromStdString(s.window)))));
``

**Comment by Myned** [comment]
Not sure why I didn't add direct matching to the search for the last window, thanks for that.

As for Windows, it still doesn't work on either of my machines after retesting. The build from https://github.com/WarmUpTill/SceneSwitcher/pull/24/commits/fabb7406bdabe87b4824aa4ef8af6cf7d7c2cdd2 works as expected being focus-only, but the build from https://github.com/WarmUpTill/SceneSwitcher/pull/24/commits/da2e11e35ebb6fd32bec2a6126e41c53fdf8c691 does not, so I can only assume it was a commit in between those. Both were tested with a single screen (laptop and desktop) using regex to match `.*Google Chrome` with a YouTube video in fullscreen. The running and focus checks are also worked as intended, it's just the fullscreen check that fails for some reason.

If you can whip up the log output, I'll gladly take a look.

**Comment by WarmUpTill** [comment]
I uploaded the build here:
https://www.dropbox.com/s/mqxiy916qambwzj/advanced-scene-switcher.dll?dl=0

Sample output:
07:16:00.222: app - b: 1048, t: -8, l: -8, r: 1928
07:16:00.222: mon - b: 1080, t: 0, l: 0, r: 1920

I think it has something to do with regex match.
As that is also not working for me, but direct match is.

I will have a look into it later today.

**Comment by Myned** [comment]
Looks like regex matching wasn't added at all, my bad. Once that builds I'll test the artifact - that was probably the only issue.

**Comment by Myned** [comment]
Well, the regex matches fine now, but I can't seem to wrap my head around the log output. The monitor size stays the same, but the window size inexplicably changes depending on fullscreen/maximization/focus, making an equality check unpredictable. At first I thought it had something to do with hidpi, but changing the scale to 100% only slightly changed the maximization offset from the monitor bounds - speaking of, the UI is squished on my laptop but not my desktop. Not sure what the issue is, and I can't increase the global window size without making other platforms extremely large.
```
05:37:42.539: User switched to scene 'Scene' // Manually switched
05:37:43.237: app - b: 1612, t: -12, l: -12, r: 2572 // No match, Chrome maximized
05:37:43.237: mon - b: 1600, t: 0, l: 0, r: 2560
05:37:48.321: app - b: 1600, t: 0, l: 0, r: 2560 // Match, YouTube fullscreen and focused
05:37:48.321: mon - b: 1600, t: 0, l: 0, r: 2560
05:37:48.334: User switched to scene 'Scene 2'
05:37:49.346: app - b: 1600, t: 0, l: 0, r: 2560
05:37:49.346: mon - b: 1600, t: 0, l: 0, r: 2560
05:37:51.390: app - b: 1599, t: 0, l: 0, r: 2560 // No match, YouTube fullscreen and not focused
05:37:51.390: mon - b: 1600, t: 0, l: 0, r: 2560
05:37:58.502: app - b: 1612, t: -12, l: -12, r: 2572 // No match, exited fullscreen and Chrome maximized
05:37:58.502: mon - b: 1600, t: 0, l: 0, r: 2560
```

**Comment by WarmUpTill** [comment]
I experience the exact same issue when testing with Google Chrome, but *only* with Google Chrome.
Other applications like notepad++, Firefox or some games do not show this behaviour.

Oddly enough the additional IsZoomed() check seems to work around this Google Chrome issue (it did not return true for other fullscreen applications) and is also working as expected for maximized windows.

Maybe you can also test once again on your end with Google Chrome once the build is done.

**Comment by Myned** [comment]
That is odd indeed. It does work for Google Chrome now, both maximized and fullscreen in YouTube, thanks for looking into it. 

There is another bug, however. Microsoft Edge (non-chromium) appears in the window list multiple times, even when the window is closed. This means that the entry itself is useless as it is always active. And since Edge removes the window title when in YouTube fullscreen (or doesn't count as a window at all, hard to tell), the fullscreen check fails. Maximize works if you specify `.+Microsoft Edge` (1 or more characters) as the title still exists then.

Given that Microsoft is phasing it out, however, I'm not sure how much we should care about this specific interaction. Though it does bring up the question of if there are other applications that behave like this.

**Comment by WarmUpTill** [comment]
Thanks for confirming it!

I do not think it is worth investigating the Microsoft Edge behavior much further, but I will also check it out on my end later today.

If there are more applications removing their window title if switched to fullscreen mode there is no way to use the window title switches successfully.
Users must then fall back to using the executable match.
But so far I have not heard any complaints about such cases, so I would assume they are very rare.

**Comment by Myned** [comment]
Understood. In that case, there isn't much more I have to add, so it's just stability testing at this point. I don't use Windows often, so I can't offer much in that regard. Up to you whether to merge as-is or wait a bit for something to pop up.

Could also offer a beta release, considering the sheer amount of changes in this PR, though users can always use a previous version if something isn't working as expected.

**Comment by WarmUpTill** [comment]
I will probably make a new release with all of these changes.
As you said, if issues arise the previous version is still available.

Thanks again!

**Comment by Myned** [comment]
Of course! Just let me know if I can help should any issues arise with the changes.
